tmail-backend/tmail-third-party/ai-bot/src/main/java/com/linagora/tmail/mailet/rag/RagListener.java

```diff
@@ -147,34 +152,79 @@ public Publisher<Void> reactiveEvent(Event event) {
         return asRagLearnableContent(messageResult)
             .doOnSuccess(text -> LOGGER.debug("RAG Listener successfully processed mailContent ***** \n{}\n *****", new DocumentId(messageResult.getMessageId())))
-            .flatMap(content -> openRagHttpClient
-                .addDocument(
+            .flatMap(content -> computeMetaData(addedEvent, messageResult, session)
+                .flatMap(metaData ->
+                    openRagHttpClient.addDocument(
                     partitionFactory.forUsername(addedEvent.getUsername()),
                     new DocumentId(messageResult.getMessageId()),
                     content,
-                    computeMetaData(addedEvent, messageResult)));
+                    metaData)));
     }
 
-    private Map<String, String> computeMetaData(MailboxEvents.Added addedEvent, MessageResult messageResult) {
+    private Mono<Map<String, String>> computeMetaData(MailboxEvents.Added addedEvent, MessageResult messageResult, MailboxSession session) {
         try {
             Message mimeMessage = parseMessage(messageResult.getFullContent().getInputStream());
             String text = new MessageContentExtractor()
                 .extract(mimeMessage)
                 .getTextBody()
                 .orElse("");
-            return  Map.of("email.subject", mimeMessage.getSubject(),
-                "email.date", DateTimeFormatter.ISO_INSTANT.format(mimeMessage.getDate().toInstant()),
-                "email.threadId", messageResult.getThreadId().serialize(),
-                "email.doctype", "com.linagora.email",
-                "email.preview", Preview.compute(text).getValue());
+
+            return getEmailIdFromMessageId(messageResult, session, getInReplyTo(mimeMessage))
+                .map(messageId -> messageId.serialize())
+                .defaultIfEmpty("")
+                .map(parentId -> Map.of(
+                    "email.subject", mimeMessage.getSubject(),
+                    "datetime", DateTimeFormatter.ISO_INSTANT.format(mimeMessage.getDate().toInstant()),
+                    "parent_id", parentId,
+                    "relationship_id", messageResult.getThreadId().serialize(),
+                    "doctype", "com.linagora.email",
+                    "email.preview", Preview.compute(text).getValue()));
         } catch (Exception e) {
             throw new RuntimeException(e);
         }
     }
 
+    private Mono<MessageId> getEmailIdFromMessageId(MessageResult messageResult, MailboxSession session, String inReplyTo) {
+        if (inReplyTo.isEmpty()) {
+            return Mono.empty();
+        }
+        return (threadIdGuessingAlgorithm.getMessageIdsInThread(
+            messageResult.getThreadId(), session).collectList().flatMapMany(
+            messageIds -> {
+                return messageIdManager.getMessagesReactive(messageIds, FetchGroup.HEADERS, session);
+            })
```

Comment:
```suggestion
        return threadIdGuessingAlgorithm.getMessageIdsInThread(messageResult.getThreadId(), session)
            .collectList()
            .flatMapMany(messageIds -> messageIdManager.getMessagesReactive(messageIds, FetchGroup.HEADERS, session))
```

Sorry code style matters to me:

 - avoid uneededbrackets
 - one operation per line and one line per operation
 - leverage lambdas